src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Metadata/DefaultModelMetadata.cs

```diff
@@ -427,6 +430,94 @@ public override bool ValidateChildren
             }
         }
 
+        /// <inheritdoc />
+        public override bool IsValidationRequired
+        {
+            get
+            {
+                if (!_isValidationRequired.HasValue)
+                {
+                    CalculateIsValidationRequired();
+                }
+
+                Debug.Assert(_isValidationRequired.HasValue);
+                return _isValidationRequired.Value;
+            }
+        }
+
+        private void CalculateIsValidationRequired()
+        {
+            var visited = new HashSet<ModelMetadata>();
+
+            if (TryVisit(this, out var requiresValidation))
+            {
+                _isValidationRequired = requiresValidation;
+            }
+            else
+            {
+                // Assume true by default.
+                _isValidationRequired = true;
+            }
+
+            bool TryVisit(ModelMetadata metadata, out bool result)
+            {
+                RuntimeHelpers.EnsureSufficientExecutionStack();
+
+                if (!visited.Add(metadata))
+                {
+                    // A cycle. Return the value as is without attempting to calculate it now. We'll calculate it at a later point.
+                    result = default;
+                    return false;
+                }
+
+                if (!(metadata is DefaultModelMetadata defaultModelMetadata))
```

Comment:
Is `DefaultModelMetadata` sealed or internal? Otherwise this should check for exact type.